### PR TITLE
add stories and component definitions

### PR DESCRIPTION
**Pull request by Kargathia**
* UI blocks stories
* UI data viewing stories
* UI blocks PlantUML diagram
* Design argumentation service layer
* Service layer PlantUML diagrams

### CODE REVIEW

File: tech_specs/subprojects.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+* Functionality is logically spread over multiple loosely coupled parts:
+    * The controller firmware (C++)
+    * The controller <-> service communication layer - multiple variants (Python)
+    * (Future) plugins for external hardware not routed through the controller (Python)
+    * Various service-side functionality such as object / measurement databases (Pythyon)
+    * The Server-Side Events API (Python)
+    * The REST API (Python)
+    * The GUI (HTML/JS using the Vue/Vuex framework)
+    * (Future) Application wrappers around the GUI, making it a mobile or desktop app.
+* The long-term vision for the product includes it evolving into a generic industrial controller - without a brewery in sight.
```
Comment (elcojacobs) [inline tech_specs/subprojects.md]: I would rephrase this into: keeping the platform open to other domains besides brewing. Do not get brewing specifics in layers that can stay generic.

File: tech_specs/subprojects.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+* The backend plugins (controlbox, legacy, and external) register endpoint handlers with the REST API.
+* The backend plugins publish events to the application event handler.
+* The logging database (InfluxDB) registers endpoints with the REST API.
+* InfluxDB subscribes to events published to the event handler.
+* SSE registers endpoints with the REST API.
+* SSE subscribes to events published to the event handler.
+
+REST, SSE, and the event handler are core components of the service layer: everything else can assume these are present.
+
+InfluxDB has no implementation-specific dependencies, and will simply log whatever is given to him. Data input is as simple as logging all published events. REST requests for data queries can also be directly forwarded, without needing an abstraction layer.
```
Comment (elcojacobs) [inline tech_specs/subprojects.md]: Maybe, but influx does not know about the mapping of its keys to activities or brews (a batch of beer going through mashing, fermentation, conditioning). A rest API could provide this and fetch the data from influx. We discussed this as a reporting module, where raw influx data is combined and grouped into reports that make sense to the user.

File: library_research/gateway_options.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+* service health checking
+
+### Nonfunctional requirements
+* easy to set up
+* easy to configure
+* easy to debug
+* good examples
+* lightweight
+* works equally well in and outside docker context
+* scalability
```
[the remark this answers is not in the document]
Comment (Kargathia) [reply library_research/gateway_options.md]: Added

File: library_research/gateway_options.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+Supports every single requirement.
+
+### [Istio](https://istio.io/)
+
+Seems capable, but documentation needs some more detailed attention.
+
+### [Tyk](https://tyk.io)
+
+* Free up to 50 000 API calls per day.
+* Only hosts the gateway, not the services
```
[the remark this answers is not in the document]
Comment (Kargathia) [reply library_research/gateway_options.md]: Janus hosts nothing. Tyk sells them hosting a gateway to services you host.

File: library_research/gateway_options.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+* support encrypted traffic.
+* reverse proxy calls to known services
+* configuration hot reload
+* support HTTP communication
+* free (as in beer)
+* actively maintained
+
+### Could haves
+* support multiple instances of services (load balancing)
+* service auto-discovery
```
[the remark this answers is not in the document]
Comment (Kargathia) [reply library_research/gateway_options.md]: updated `configuration hot reload` to include "or runtime registration"

File: stories/block_stories.feature [hunk 6]
```diff
@@ -0,0 +1,67 @@
+# Note: alternative name for "UI block" could be "Block"
+
+Feature: UI blocks
+    UI blocks represent distinct hardware functionality
```
[the remark this answers is not in the document]
Comment (Kargathia) [reply stories/block_stories.feature]: Added that UI block is sub-category of widget.

File: stories/block_stories.feature [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    UI blocks can be associated with one or more hardware elements
+    UI blocks do not have to be associated with hardware elements
+    UI blocks can be associated with dummy elements
+    UI blocks can be associated with simulation elements
+    UI blocks can change their association
+    UI blocks should match user expectation of having
```
[the remark this answers is not in the document]
Comment (Kargathia) [reply stories/block_stories.feature]: ... I have no idea. removed it.

File: stories/block_stories.feature [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+Feature: UI blocks
+    UI blocks represent distinct hardware functionality
+    UI blocks can be associated with one or more hardware elements
+    UI blocks do not have to be associated with hardware elements
+    UI blocks can be associated with dummy elements
+    UI blocks can be associated with simulation elements
+    UI blocks can change their association
+    UI blocks should match user expectation of having 
+    UI blocks can be a grouping of other UI blocks
```
[the remark this answers is not in the document]
Comment (Kargathia) [reply stories/block_stories.feature]: That seems the more complex alternative.

File: stories/block_stories.feature [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    UI blocks have a unique identifier independent of association
+    UI blocks are grouped in one or more processes
+
+Scenario: Add UI block from UI blockbox
+    Given there is an active process
+    When I add a new UI block from the UI blockbox
```
[the remark this answers is not in the document]
Comment (Kargathia) [reply stories/block_stories.feature]: Indeed a better name. Fixed.

File: stories/block_stories.feature [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    Then a new UI block is added to the process
+    And the new UI block is associated with a dummy element
+    And the new UI block has a unique identifier
+
```
[the remark this answers is not in the document]
Comment (Kargathia) [reply stories/block_stories.feature]: Reworking overall hierarchy to `control center` -> `dashboard` -> `view` -> `widget`

File: stories/block_stories.feature [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    When I press the UI block Wizard button
+    Then I can select an association for the UI block
+
+Scenario: Autodeploy UI block configuration
+    Given there are unassociated UI blocks
+    And I have plugged corresponding physical objects into the controller
+    And I have not configured the controller
+    When I press the Autodeploy Blocks button
+    Then the application configures all new objects on the controller
+    And the application attempts to associate all UI blocks
```
[the remark this answers is not in the document]
Comment (Kargathia) [reply stories/block_stories.feature]: Fixed.

File: stories/block_stories.feature [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    And the process contains a UI block
+    When I press the Associate Hardware button
+    Then I am asked to plug in the associated hardware component
+    And the UI block is associated with the hardware component
+
+Scenario: Assign UI block simulator association
+    Given there is an active process
+    And the process contains a UI block
+    When I press the Associate Simulator button
+    Then the UI block is associated with a simulator component
```
[the remark this answers is not in the document]
Comment (Kargathia) [reply stories/block_stories.feature]: Fixed.

File: stories/block_stories.feature [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+Scenario: Assign UI block-to-UI block association
```
[the remark this answers is not in the document]
Comment (Kargathia) [reply stories/block_stories.feature]: Renaming to `UI block grouping association`

File: tech_specs/eventbus.md [hunk 14]
```diff
@@ -0,0 +1,32 @@
+# Event handling
```
[the remark this answers is not in the document]
Comment (Kargathia) [reply tech_specs/eventbus.md]: Added conclusions and implementation.

File: tech_specs/service_components.txt [hunk 15]
```diff
@@ -0,0 +1,111 @@
+@startuml "Service Components"
```
[the remark this answers is not in the document]
Comment (Kargathia) [reply tech_specs/service_components.txt]: Seems to be the case. I'll split the diagrams, as online viewing of diagrams is a desirable functionality.

File: tech_specs/subprojects.md (hunk 1 above)
Comment (Kargathia) [reply tech_specs/subprojects.md]: Fixed

File: tech_specs/subprojects.md (hunk 2 above)
Comment (Kargathia) [reply tech_specs/subprojects.md]: Described under `Splitting up` chapter, as the concept was introduced here.

File: stories/data_stories.feature [hunk 16]
```diff
@@ -1,34 +1,34 @@
...
+Configuration changes are captured
+Events are captured
+User annotations are captured
 
-Scenario: View process data
-    Given there is an active process
-    And the active process contains UI blocks with a hardware association
-    When I click the View Process Data button
-    Then I should see all data generated by the process UI blocks
+    Scenario: Display process data
```
Comment (elcojacobs) [inline stories/data_stories.feature]: Views get live updates automatically, no clicking should be needed outside of navigating to a view.

File: tech_specs/component_definitions.md [hunk 17]
```diff
@@ -25,24 +25,32 @@ In order to make these ends meet, we can identify the following levels:
...
+The top level grouping inside the GUI application. All sensors, actuators, and settings for a single service are somewhere in here. UI blocks are stored here, so multiple views can share them.
 
-### Dashboard (eg. "Fermentation" / "Overview" / "All temperature sensors")
+### View (eg. "Fermentation" / "Overview" / "All temperature sensors")
 
-A clearly defined subset of an activity. This involves a subset of all UI blocks, but generally uses all assigned UI blocks simultaneously.
+A canvas displaying one or more widgets. Views do not directly display data or actions: they contain all objects (widgets) that do.
 
-It should be possible that multiple dashboards share physical UI blocks.
+Views can contain other views. Views can share widgets.
```
Comment (elcojacobs) [inline tech_specs/component_definitions.md]: Not sure about this. It will probably be easier if we define widgets inside a view.
If 2 views want to display the same widget, it is defined in each. They map to the same hardware/service data, but this is separate from the view widget.

File: tech_specs/component_definitions.md [hunk 18]
```diff
@@ -25,24 +25,32 @@ In order to make these ends meet, we can identify the following levels:
...
+### Widget (eg. UI Block / Graph)
+
+All discrete objects in the UI are widgets. UI blocks display state and actions, graphs display history. Widgets can be created and moved by users.
+
+The same widget can be present on multiple views.
 
 ### UI block (eg. PID, actuator, mutex)
 
-Performs a specific, abstract task. A UI block isn't busy brewing beer. It might keep temperature at a certain level, or even just toggle the heater.
+Represents a hardware component performing a specific, abstract task. A UI block isn't busy brewing beer. It might keep temperature at a certain level, or even just toggle the heater.
```
Comment (elcojacobs) [inline tech_specs/component_definitions.md]: As said before: the view block only represents a task. It doesn't perform any task, so it also doesn't keep temperature. It is a stateless and actionless display block. It can only display data or emit events.  If it represents a PID, the PID block in the firmware maintains the temp. They are linked in a separate step, so we should keep them separate. View block <--> active controller block.

File: uml/microservices_diagram.txt [hunk 19]
```diff
@@ -0,0 +1,40 @@
+@startuml "Micro services"
```
Comment (elcojacobs) [inline uml/microservices_diagram.txt]: Looks good!
